File: base/001/solver.cpp
#include <iostream>
#include <sstream>
#include <cstdio>
using namespace std;

struct Calc{
    int bateria;
    int limiteBateria;
    float display;

    Calc(int limiteBateria = 0){
        this->limiteBateria = limiteBateria;
        this->bateria = 0;
        this->display = 0.0;
    }
    void charge(int carga){
        bateria += carga;
        if(bateria > limiteBateria)
            bateria = limiteBateria;
    }
    bool gastarBateria(){
        if(bateria == 0){
            cout << "fail: bateria insuficiente" << endl;
            return false;
        }
        this->bateria -= 1;
        return true;
    }
    void soma(int a, int b){
        if(gastarBateria())
            display = a + b;
    }
    void div(int a, int b){
        if(!gastarBateria())
            return;
        if(b != 0)
            display = (float) a / b;
        else
            cout << "fail: divisao por zero" << endl;        
    }
    void show(){
        printf("display = %.2f, battery = %d\n", this->display, this->bateria);
    }
};

int main(){
    Calc calc;
    string line;
    while(true){
        getline(cin, line);
        cout << "$" << line << endl;
        if(line == "end")
            break;

        stringstream ss(line);
        string op;
        ss >> op;
        if(op == "help"){
            cout << "sum _a _b; div _a _b; show; charge _carga; end";
        }else if(op == "init"){
            int max;
            ss >> max;
            calc = Calc(max);
        }else if(op == "sum"){
            int a, b;
            ss >> a >> b;
            calc.soma(a, b);
        }else if(op == "div"){
            int a, b;
            ss >> a >> b;
            calc.div(a, b);
        }else if(op == "charge"){
            int charge;
            ss >> charge;
            calc.charge(charge);
        }else if(op == "show"){
            calc.show();
        }else 
            cout << "fail: comando invalido" << endl;
    }
    return 0;
}

File: base/001/Solver.java
import java.util.Scanner;
import java.text.*;

class Calculadora {
	int bateriaMaxima;
	int bateria;
	float display;
	
	public Calculadora(int bateriaMaxima) {
		this.bateriaMaxima = bateriaMaxima;
		this.bateria = 0;
		this.display = 0.0f;
	}
	
	void charge(int value){
		if(value < 0)
			return;
		this.bateria += value;
		if(this.bateria > this.bateriaMaxima)
			this.bateria = this.bateriaMaxima;
	}
	boolean gastarBateria() {
		if(this.bateria == 0){
			System.out.println("fail: bateria insuficiente");
			return false;
		}
		this.bateria -= 1;
		return true;
	}
	
	void soma(int a, int b){
		if(gastarBateria())
			this.display = (a + b);
	}
	
	void div(int num, int den){
		if(!gastarBateria())
			return;
		if(den == 0){
			System.out.println("fail: divisao por zero");
		}
		else
			this.display = (float) num / den;
	}
	
	public String toString(){
		DecimalFormat form = new DecimalFormat("0.00");
		return "display = " + form.format(this.display) + ", battery = " + this.bateria;
	}
}

class Solver{
	
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		Calculadora calc = new Calculadora(0);
		while(true){
			String line = scan.nextLine();
			System.out.println("$" + line);
			if(line.equals("end"))
				break;
				String ui[] = line.split(" ");
				if(ui[0].equals("help"))
					System.out.println("soma _a _b; show; div _a _b; charge _value");
				else if(ui[0].equals("init"))
					calc = new Calculadora(Integer.parseInt(ui[1]));
				else if(ui[0].equals("show"))
					System.out.println(calc);
				else if(ui[0].equals("charge")){
					calc.charge(Integer.parseInt(ui[1]));
				}else if(ui[0].equals("sum")){
					calc.soma(Integer.parseInt(ui[1]), Integer.parseInt(ui[2]));
				}else if(ui[0].equals("div")){
					calc.div(Integer.parseInt(ui[1]), Integer.parseInt(ui[2]));
				}else
					System.out.println("fail: comando invalido");
		}
		scan.close();
	}
}